### ISSUE TITLE
[BUG] Sequilize foreign key duplication bug

### ISSUE THREAD

**Issue by HopefulFire** [sequelize/sequelize#13330]
## Issue Description
Keys seem to duplicate in sequelize ORM, with different case sensitivity. For example, if I define `userId`, the query is for `userId`, but also it looks for `UserId`. Then it gets mad that the column `UserId` doesn't exist (because I never migrated the database to have that column. I am extremely frustrated and this is causing delays.

### What are you doing?

I cannot really provide a SSCCE since this involves migrations. I can, however, share some code:

- `migrations/2021...-create-jwt-whitelist.js`
```javascript
"use strict";
module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable("JwtWhitelists", {
      id: {
        allowNull: false,
        autoIncrement: true,
        primaryKey: true,
        type: Sequelize.BIGINT,
      },
      jwt: {
        allowNull: false,
        type: Sequelize.TEXT,
        unique: true,
      },
      userId: {
        allowNull: false,
        type: Sequelize.BIGINT,
        unique: true,
      },
      createdAt: {
        allowNull: false,
        type: Sequelize.DATE,
      },
      updatedAt: {
        allowNull: false,
        type: Sequelize.DATE,
      },
    });
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable("JwtWhitelists");
  },
};
```

- `migrations/2021...-create-user.js`
```javascript
"use strict";
module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable("Users", {
      id: {
        allowNull: false,
        autoIncrement: true,
        primaryKey: true,
        type: Sequelize.BIGINT,
      },
      email: {
        type: Sequelize.TEXT,
      },
      passwordDigest: {
        type: Sequelize.TEXT,
      },
      emailConfirmed: {
        type: Sequelize.BOOLEAN,
      },
      isAdmin: {
        type: Sequelize.BOOLEAN,
      },
      createdAt: {
        allowNull: false,
        type: Sequelize.DATE,
      },
      updatedAt: {
        allowNull: false,
        type: Sequelize.DATE,
      },
    });
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable("Users");
  },
};

```

- `models/jwtwhitelist.js`
```javascript
"use strict";
const { Model } = require("sequelize");
module.exports = (sequelize, DataTypes) => {
  class JwtWhitelist extends Model {
    /**
     * Helper method for defining associations.
     * This method is not a part of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    static associate(models) {
      JwtWhitelist.belongsTo(models.User);
    }
  }
  JwtWhitelist.init(
    {
      jwt: DataTypes.TEXT,
      userId: DataTypes.BIGINT,
    },
    {
      sequelize,
      modelName: "JwtWhitelist",
    }
  );
  return JwtWhitelist;
};
```

- `models/user.js`
```javascript
"use strict";
const { Model } = require("sequelize");
module.exports = (sequelize, DataTypes) => {
  class User extends Model {
    /**
     * Helper method for defining associations.
     * This method is not a part of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    static associate(models) {
      User.hasOne(models.JwtWhitelist);
    }
  }
  User.init(
    {
      email: DataTypes.TEXT,
      passwordDigest: DataTypes.TEXT,
      emailConfirmed: DataTypes.BOOLEAN,
      isAdmin: DataTypes.BOOLEAN,
    },
    {
      sequelize,
      modelName: "User",
    }
  );
  return User;
};
```

 

### What do you expect to happen?

I wanted the query to return a jwtwhitelist associated with the user, or return null if none was found.

### What is actually happening?

The output is some nonsense, here I will quote it: `RETURNING "id","jwt","userId","createdAt","updatedAt","UserId";
2021-06-18 23:11:00.813-nodemon DatabaseError [SequelizeDatabaseError]: column "UserId" does not exist`

```
2021-06-18 23:11:00.471-nodemon POST /api/authentication/login -> authentication#login
2021-06-18 23:11:00.724-nodemon Executing (default): SELECT "id", "email", "passwordDigest", "emailConfirmed", "isAdmin", "createdAt", "updatedAt" FROM "Users" AS "User" WHERE "User"."email" = 'jadelynn@masker.family' LIMIT 1;
2021-06-18 23:11:00.786-nodemon Executing (default): SELECT "key" FROM "SigningKeys" AS "SigningKey" ORDER BY "SigningKey"."createdAt" DESC LIMIT 1;
2021-06-18 23:11:00.788-nodemon Signing key not found, generating new one!
2021-06-18 23:11:00.795-nodemon Executing (default): INSERT INTO "SigningKeys" ("id","key","createdAt","updatedAt") VALUES (DEFAULT,$1,$2,$3) RETURNING "id","key","createdAt","updatedAt";
2021-06-18 23:11:00.804-nodemon Executing (default): DELETE FROM "JwtWhitelists" WHERE "userId" = '1'
2021-06-18 23:11:00.808-nodemon Executing (default): INSERT INTO "JwtWhitelists" ("id","userId","createdAt","updatedAt") VALUES (DEFAULT,$1,$2,$3) RETURNING "id","jwt","userId","createdAt","updatedAt","UserId";
2021-06-18 23:11:00.813-nodemon DatabaseError [SequelizeDatabaseError]: column "UserId" does not exist
2021-06-18 23:11:00.813-nodemon     at Query.formatError (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/dialects/postgres/query.js:386:16)
2021-06-18 23:11:00.813-nodemon     at Query.run (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/dialects/postgres/query.js:87:18)
2021-06-18 23:11:00.813-nodemon     at processTicksAndRejections (internal/process/task_queues.js:95:5)
2021-06-18 23:11:00.813-nodemon     at async /home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/sequelize.js:619:16
2021-06-18 23:11:00.813-nodemon     at async PostgresQueryInterface.insert (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/dialects/abstract/query-interface.js:749:21)
2021-06-18 23:11:00.813-nodemon     at async JwtWhitelist.save (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/model.js:3954:35)
2021-06-18 23:11:00.813-nodemon     at async Function.create (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/sequelize/lib/model.js:2207:12)
2021-06-18 23:11:00.813-nodemon     at async Promise.all (index 0) {
2021-06-18 23:11:00.813-nodemon   parent: error: column "UserId" does not exist
2021-06-18 23:11:00.813-nodemon       at Parser.parseErrorMessage (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:287:98)
2021-06-18 23:11:00.813-nodemon       at Parser.handlePacket (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:126:29)
2021-06-18 23:11:00.813-nodemon       at Parser.parse (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:39:38)
2021-06-18 23:11:00.813-nodemon       at Socket.<anonymous> (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/index.js:11:42)
2021-06-18 23:11:00.813-nodemon       at Socket.emit (events.js:375:28)
2021-06-18 23:11:00.813-nodemon       at addChunk (internal/streams/readable.js:290:12)
2021-06-18 23:11:00.813-nodemon       at readableAddChunk (internal/streams/readable.js:265:9)
2021-06-18 23:11:00.813-nodemon       at Socket.Readable.push (internal/streams/readable.js:204:10)
2021-06-18 23:11:00.813-nodemon       at TCP.onStreamRead (internal/stream_base_commons.js:188:23) {
2021-06-18 23:11:00.813-nodemon     length: 174,
2021-06-18 23:11:00.813-nodemon     severity: 'ERROR',
2021-06-18 23:11:00.813-nodemon     code: '42703',
2021-06-18 23:11:00.813-nodemon     detail: undefined,
2021-06-18 23:11:00.813-nodemon     hint: 'Perhaps you meant to reference the column "JwtWhitelists.userId".',
2021-06-18 23:11:00.813-nodemon     position: '149',
2021-06-18 23:11:00.813-nodemon     internalPosition: undefined,
2021-06-18 23:11:00.813-nodemon     internalQuery: undefined,
2021-06-18 23:11:00.813-nodemon     where: undefined,
2021-06-18 23:11:00.813-nodemon     schema: undefined,
2021-06-18 23:11:00.813-nodemon     table: undefined,
2021-06-18 23:11:00.813-nodemon     column: undefined,
2021-06-18 23:11:00.813-nodemon     dataType: undefined,
2021-06-18 23:11:00.813-nodemon     constraint: undefined,
2021-06-18 23:11:00.813-nodemon     file: 'parse_relation.c',
2021-06-18 23:11:00.813-nodemon     line: '3504',
2021-06-18 23:11:00.813-nodemon     routine: 'errorMissingColumn',
2021-06-18 23:11:00.813-nodemon     sql: 'INSERT INTO "JwtWhitelists" ("id","userId","createdAt","updatedAt") VALUES (DEFAULT,$1,$2,$3) RETURNING "id","jwt","userId","createdAt","updatedAt","UserId";',
2021-06-18 23:11:00.813-nodemon     parameters: [
2021-06-18 23:11:00.813-nodemon       '1',
2021-06-18 23:11:00.813-nodemon       '2021-06-19 03:11:00.805 +00:00',
2021-06-18 23:11:00.813-nodemon       '2021-06-19 03:11:00.805 +00:00'
2021-06-18 23:11:00.813-nodemon     ]
2021-06-18 23:11:00.813-nodemon   },
2021-06-18 23:11:00.813-nodemon   original: error: column "UserId" does not exist
2021-06-18 23:11:00.813-nodemon       at Parser.parseErrorMessage (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:287:98)
2021-06-18 23:11:00.813-nodemon       at Parser.handlePacket (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:126:29)
2021-06-18 23:11:00.813-nodemon       at Parser.parse (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/parser.js:39:38)
2021-06-18 23:11:00.813-nodemon       at Socket.<anonymous> (/home/jade/code/AracelliVPhotography/aracelli-v-photography-website/node_modules/pg-protocol/dist/index.js:11:42)
2021-06-18 23:11:00.813-nodemon       at Socket.emit (events.js:375:28)
2021-06-18 23:11:00.813-nodemon       at addChunk (internal/streams/readable.js:290:12)
2021-06-18 23:11:00.813-nodemon       at readableAddChunk (internal/streams/readable.js:265:9)
2021-06-18 23:11:00.813-nodemon       at Socket.Readable.push (internal/streams/readable.js:204:10)
2021-06-18 23:11:00.813-nodemon       at TCP.onStreamRead (internal/stream_base_commons.js:188:23) {
2021-06-18 23:11:00.813-nodemon     length: 174,
2021-06-18 23:11:00.813-nodemon     severity: 'ERROR',
2021-06-18 23:11:00.813-nodemon     code: '42703',
2021-06-18 23:11:00.813-nodemon     detail: undefined,
2021-06-18 23:11:00.813-nodemon     hint: 'Perhaps you meant to reference the column "JwtWhitelists.userId".',
2021-06-18 23:11:00.813-nodemon     position: '149',
2021-06-18 23:11:00.813-nodemon     internalPosition: undefined,
2021-06-18 23:11:00.813-nodemon     internalQuery: undefined,
2021-06-18 23:11:00.813-nodemon     where: undefined,
2021-06-18 23:11:00.813-nodemon     schema: undefined,
2021-06-18 23:11:00.813-nodemon     table: undefined,
2021-06-18 23:11:00.813-nodemon     column: undefined,
2021-06-18 23:11:00.813-nodemon     dataType: undefined,
2021-06-18 23:11:00.813-nodemon     constraint: undefined,
2021-06-18 23:11:00.813-nodemon     file: 'parse_relation.c',
2021-06-18 23:11:00.813-nodemon     line: '3504',
2021-06-18 23:11:00.813-nodemon     routine: 'errorMissingColumn',
2021-06-18 23:11:00.813-nodemon     sql: 'INSERT INTO "JwtWhitelists" ("id","userId","createdAt","updatedAt") VALUES (DEFAULT,$1,$2,$3) RETURNING "id","jwt","userId","createdAt","updatedAt","UserId";',
2021-06-18 23:11:00.813-nodemon     parameters: [
2021-06-18 23:11:00.813-nodemon       '1',
2021-06-18 23:11:00.813-nodemon       '2021-06-19 03:11:00.805 +00:00',
2021-06-18 23:11:00.813-nodemon       '2021-06-19 03:11:00.805 +00:00'
2021-06-18 23:11:00.813-nodemon     ]
2021-06-18 23:11:00.813-nodemon   },
2021-06-18 23:11:00.813-nodemon   sql: 'INSERT INTO "JwtWhitelists" ("id","userId","createdAt","updatedAt") VALUES (DEFAULT,$1,$2,$3) RETURNING "id","jwt","userId","createdAt","updatedAt","UserId";',
2021-06-18 23:11:00.813-nodemon   parameters: [
2021-06-18 23:11:00.813-nodemon     '1',
2021-06-18 23:11:00.813-nodemon     '2021-06-19 03:11:00.805 +00:00',
2021-06-18 23:11:00.813-nodemon     '2021-06-19 03:11:00.805 +00:00'
2021-06-18 23:11:00.813-nodemon   ]
2021-06-18 23:11:00.813-nodemon }
2021-06-18 23:11:00.816-nodemon Status Code: 400


```

### Environment

- Sequelize version: sequelize@6.6.2 
- Node.js version: v14.17.1 
- Operating System: Ubuntu Linux
- If TypeScript related: TypeScript version: n/a

### How does this problem relate to dialects?

- [ ] I think this problem happens regardless of the dialect.
- [ ] I think this problem happens only for the following dialect(s): 
-  [x] I don't know, I was using POSTGRESQL, with connector library version pg@8.6.0 and database version Ubuntu 13.3-1.pgdg20.04+1 (however I think it is dialect related as not all databases are case sensitive)

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
-  [x] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by JacobLey** [comment]
`JwtWhitelist.belongsTo(models.User);` -> `JwtWhitelist.belongsTo(models.User, { foreignKey: 'userId' });`
`User.hasOne(models.JwtWhitelist);` -> `User.hasOne(models.JwtWhitelist, { foreignKey: 'userId' });`

The `UserId` is coming from these association declarations, that are using the default name (`model.name + 'id'`). Declare the name manually to overlap your column naming and it should work as intended

**Comment by HopefulFire** [comment]
Thank you for your reply!
I tried renaming the columns to UserId, if that's what you mean. And I checked the database, the columns UserId do not exist.
That's what the problem is, it's creating a column in the ORM but not in the database. 
I would personally prefer it if migrations were separate from associations anyway, as issues like this can arise. I actually originally started with declaring the columns userId, then rolled back the database and tried UserId, then rolled back the database again and tried without any userId. Every time was the same result: a duplicate UserId column would exist for every userId column.
I am rebuilding the application from scratch, so I will let you know if the issue persists or if it continues.

**Comment by HopefulFire** [comment]
More on the separation of concerns:
I come from a ruby on rails background, where model associations have nothing to do with database declarations. If the required association doesn't exist, like a join table linking friends to friends, it will simply create an error.
I think that is a better way of doing things than manually creating tables from the models

**Comment by AdamAld** [comment]
I'd recommend the migration order of `Users` -> `JwtWhitelist`, with the latter containing the relation directly in the migration.
To add to the explicitness outlined by @JacobLey , it also helps to include the `field` option in your migration and model files

```javascript 
      userId: {
        allowNull: false,
        type: Sequelize.BIGINT,
        unique: true,
        references: { // Explicitly tells Sequelize to create a foreign key relation with `Users`.`id`
          model: 'Users',
          key: 'id',
        },
        field: 'userId', // Explicitly tells Sequelize the database field name
      },
```
Additionally, It would be beneficial to give the [naming-strategies](https://sequelize.org/master/manual/naming-strategies.html) documentation a read through.

I hope this helps!

**Comment by HopefulFire** [comment]
That is helpful.
I don't think Sequelize should create UserId just because I say `User.hasOne(models.JwtWhitelist);`, but I can accept it if there is a way around it. Also, I will definitly use `underscored: true`!
I apologize if I seem unknowlegable, I am very new to Software Engineering. I only recently have learned what a reference is in postgresql, for instance. I still think there is a bug, but if I was not following conventions than it's my own fault I found it.
I thought this was a major issue with sequelize, as an outsider.

**Comment by adeeb1** [comment]
I'm glad I found this thread. I wanted to leave a comment that this also caused huge issues for me because the documentation is wrong. In some places it states that a camel-cased column name (ex: `userId`) is used but it clearly looks for pascal casing (ex: `UserId`).

In my situation, I had `userId` and `interestId` keys on my `UserInterests` join table. I had to define my associations like this:

```
User.belongsToMany(Interest, { through: UserInterest, foreignKey: 'userId' });
Interest.belongsToMany(User, { through: UserInterest, foreignKey: 'interestId' });

UserInterest.belongsTo(User, { foreignKey: 'userId' });
UserInterest.belongsTo(Interest, { foreignKey: 'interestId' });
```

I will file an issue to update the documentation.
